python/dgl/graphbolt/utils/sample_utils.py

```diff
@@ -0,0 +1,93 @@
+"""Utility functions for sampling."""
+
+from collections import defaultdict
+from typing import Dict, Tuple, Union
+
+import torch
+
+
+def unique_and_compact_node_pairs(
+    node_pairs: Union[
+        Tuple[torch.Tensor, torch.Tensor],
+        Dict[Tuple[str, str, str], Tuple[torch.Tensor, torch.Tensor]],
+    ]
+):
+    """
+    Compact node pairs and return unique nodes (per type).
+
+    Parameters
+    ----------
+    node_pairs : Tuple[torch.Tensor, torch.Tensor] or \
+        Dict(Tuple[str, str, str], Tuple[torch.Tensor, torch.Tensor])
+        Node pairs representing source-destination edges.
+        - If `node_pairs` is a tuple: It means the graph is homogeneous.
+        Also, it should be in the format ('u', 'v') representing source
+        and destination pairs. And IDs inside are homogeneous ids.
+        - If `node_pairs` is a dictionary: The keys should be edge type and
+        the values should be corresponding node pairs. And IDs inside are
+        heterogeneous ids.
+
+    Returns
+    -------
+    Tuple[node_pairs, unique_nodes]
+        The compacted node pairs, where node IDs are replaced with mapped node
+        IDs, and the unique nodes (per type).
+        "Compacted node pairs" indicates that the node IDs in the input node
+        pairs are replaced with mapped node IDs, where each type of node is
+        mapped to a contiguous space of IDs ranging from 0 to N.
+
+    Examples
+    --------
+    >>> import dgl.graphbolt as gb
+    >>> N1 = torch.LongTensor([1, 2, 2])
+    >>> N2 = torch.LongTensor([5, 6, 5])
+    >>> node_pairs = {("n1", "e1", "n2"): (N1, N2),
+    ...  ("n2", "e2", "n1"): (N2, N1)}
+    >>> unique_nodes, compacted_node_pairs = gb.unique_and_compact_node_pairs(
+    ... node_pairs
+    ... )
+    >>> print(unique_nodes)
+    {'n1': tensor([1, 2]), 'n2': tensor([5, 6])}
+    >>> print(compacted_node_pairs)
+    {('n1', 'e1', 'n2'): (tensor([0, 1, 1]), tensor([0, 1, 0])),
+    ('n2', 'e2', 'n1'): (tensor([0, 1, 0]), tensor([0, 1, 1]))}
+    """
+    is_heterogeneous = isinstance(node_pairs, Dict)
```

Comment:
Such that the 2 `if not is_heterogeneous` can be replaced with `if is_homogeneous`

```suggestion
    is_homogeneous = not isinstance(node_pairs, Dict)
```

Reply:
make sense, use `is_homogeneous `